### DISCUSSION TITLE
Simple way of getting location of devenv.exe

### DISCUSSION THREAD

**Discussion by tapika** [cake-build/cake#3889] [Q&A] [answered]
I wanted to build with cake filtered solution - *.slnf.

https://docs.microsoft.com/en-us/visualstudio/ide/filtered-solutions?view=vs-2019

And I think it's not possible to be done with `msbuild.exe` - I've tried to switch from `msbuild.exe` to `devenv.exe`.

And in order to do that one - I needed to know location of `devenv.exe` - and checked that there exists functions which can be used, but they are quite hidden one.

I wrote something like this:
```
Type t = typeof(MSBuildSettings).Assembly.GetType("Cake.Common.Tools.MSBuild.MSBuildResolver");
var m = t.GetMethod("GetMSBuildPath", System.Reflection.BindingFlags.Public | System.Reflection.BindingFlags.Static);
            
FilePath msbuildExePath = (FilePath)m.Invoke(null, new object[] { new FileSystem(), 
    new CakeEnvironment(new CakePlatform(), new CakeRuntime()), MSBuildToolVersion.VS2017, MSBuildPlatform.x64, null });
string dir = new FileInfo(msbuildExePath.FullPath).Directory.Parent.Parent.Parent.Parent.FullName;
string devenvPath = Path.Combine(dir, @"Common7\IDE\devenv.exe");
```

But would it be possible to simplify this - maybe making specific methods public ?

**Comment by tapika** [comment]
Just to be fully strict - apparently filtered solutions are not supported by vs2017 toolset - but they are supported by vs2019 - so theoretically replacing `MSBuildToolVersion.VS2017` to `MSBuildToolVersion.VS2019` should fix the issue.

However - while testing how `msbuild.exe` works comparing to `devenv.com` (That's right - I had also extension incorrect) - I've 
noticed that `msbuild.exe` fails where `devenv.com` works out of box.

We have special native C++ nuget package included into project (`microsoft.dxsdk.d3dx`) and for some reason C++ cannot find right include folders, which in a turn manifests as an error.

`fatal error C1083: Cannot open include file: 'd3dx9tex.h': No such file or directory`

Would be interesting to find out root cause of that, but this does not relate to this discussion itself.

I think building using `devenv.com` should be supported as well as building via `msbuild.exe`.

**Comment by jrdodds** [comment, marked as answer]
To locate installed instances of Visual Studio on Windows use the `vswhere` tool. See "[Tools for detecting and managing Visual Studio instances](https://learn.microsoft.com/en-us/visualstudio/install/tools-for-managing-visual-studio-instances?view=vs-2022)" and [microsoft/vswhere](https://github.com/microsoft/vswhere). (Note that `vswhere` is Windows only.)
